### ISSUE TITLE
tighten up dosemu security

### ISSUE THREAD

**Issue by stsp** [dosemu2/dosemu2#2181]
I am under the heavy impression of this:
https://lwn.net/ml/oss-security/CAN_LGv3GAmdpaXVCjwp1UAH_Z6KKDnqydj68Oj4jmXRwwPE=Uw@mail.gmail.com/
and am no longer comfortable with our
minimal security model which is to only
provide the safe defaults.
Now I think something really radical needs
to be done.
My current thinking is to make the dosemu.bin
binary owned by dosemu2:dosemu2 rather
than root:root, and make it setuid+setgid.
The very simple code will need to be added
to setresuid() euid back to the user for the
time of an initialization, and then do setuid()
to switch to dosemu2 permanently, and
access DOS files only with that identity.

This way we avoid dosemu2 from ever
reading any user's files after init. The
user will need to chown all DOS files to
dosemu2:dosemu2 and add himself to
dosemu2 group in order to be able to
access the DOS files as before.

The same trick can be done with
`sudo setpriv --euid=dosemu2 --egid=dosemu2`
but I think requiring the user to set up
sudoers gives more troubles than to
just use setuid+setgid.

@skitt @taviso everyone, what do you
think? Does this idea sound sensible?

**Comment by skitt** [comment]
Regular users can’t give files away like that, so anything requiring a change in ownership won’t work. Groups would work, but that also requires that users belong to the target group so it’s not great.

I wonder how much effort it would take to make dosemu2 work in a flatpak, or to have it use bubblewrap to set up its own sandboxing.

I don’t think dosemu2 is a high-value target though, so probably not worth worrying about too much ;-). (I don’t mean that improving dosemu2 security isn’t useful, just that it shouldn’t result in worry or stress.)

**Comment by stsp** [comment]
> Groups would work, but that also requires that users belong to the target group so it’s not great.

Is it a bad practice to have the
post-inst script adding the user
to the needed group?
But in either case, to get the group-only
approach to work, you still need the
dosemu2 process to change its uid
to "dosemu2" so that it no longer owns
the user's files. And that requires CAP_SETUID.
Additionally you need all the DOS files to
be group-writable, because in the case of
group-only approach dosemu2 will need
to use the group perms to access the DOS
files, rather than the owner's perms.
This will lead to tons of bugs when the
user forgot to make this or that file
group-writable. So yes, group-only approach
doesn't feel too great.

> or to have it use bubblewrap to set up its own sandboxing.

Could you please explain in 2 words
how exactly you see that helping?
The ultimate goal is to allow dosemu2
to access the DOS files, but not the
user's files. In that task I can only
imagine 2 solutions: either you give
away the ownership of DOS files, or
you use chroot. I don't think bubblewrap
adds any "third" solution, or does it?

> I don’t think dosemu2 is a high-value target though, so probably not worth worrying about too much ;-).

That's my thinking as well, and we've
yet to see some DOS virus hacking
dosbox (a much more widely-available
target than dosemu2). But still...

**Comment by stsp** [comment]
Ok so bubblewrap uses the chroot-alike
technique with the mount namespaces.
Maybe this is indeed more promising
than the UID-based separation, RTFMing... :)

**Comment by skitt** [comment]
Right, I’m thinking that a useful way of using something like bubblewrap would be to process mounts and then deny access to anything outside those; for example, once configuration files are processed, no further mounts can be added. This would be annoying for some users so it would be nice if it could be disabled...

**Comment by stsp** [comment]
> Right, I’m thinking that a useful way of using something like bubblewrap would be to process mounts and then deny
> access to anything outside those; for example, once configuration files are processed, no further mounts can be added.

How do you do that with bubblewrap?
So far the examples I've seen, were
passing the needed paths to `bwrap`,
which means you mount things even
before starting the process.

> This would be annoying for some users so it would be nice if it could be disabled...

No because this is already so for a long
time. You can't mount arbitrary dir with
`lredir` (which is why lredir was renamed
multiple times - it simply doesn't do what
it did before). Likewise you can't execute
arbitrary binary with unix.com, and so on.
We have safe defaults, strong restrictions
and good sandboxing (unless you enable
native DPMI - its not yet sandboxed, but
who cares if its not default).

So why to bother? Well, mostly because
we have a dosemu1-inherited code like
8287579. Do you have any idea why the
overflow check was inverted, and why the
surrounding code is so heavily obfuscated?
Maybe just a coincidence, maybe not.
ALso I wouldn't trust the dosemu2-written
code much as well:
```
$ git diff -r master |diffstat
...
629 files changed, 52147 insertions(+), 19077 deletions(-)
```
And this doesn't account for fdpp,
comcom64, dj64dev. The amount of
added code is huge, the amount of
developers and reviewers is tiny. The
only way to write as much code, is to
use the dirty rapid prototyping, or
nothing would be written at all.
I am sure other emulators have smaller
diffstats with more developers - they
have better code, something we can't
afford.

**Comment by andrewbird** [comment]
Perhaps it would help to setup Coverity static scanning from Github?

**Comment by stsp** [comment]
Why not?

**Comment by andrewbird** [comment]
I did set it up once for dosemu once, but it will take quite some effort to make dosemu pass the checks cleanly

**Comment by stsp** [comment]
We probably don't need to pass them
cleanly? The stats would be enough, so
that the errors are not added up.
Eventually we'll get them fixed.

**Comment by stsp** [comment]
I started to experiment with user
namespaces (which are needed together
with mount namespaces here), and I
don't understand why kernel allows to
only write 1 extent (string) to `/proc/self/uid_map`.
I would like to create 2 mappings: for root
and unprivileged user within the namespace,
mapping to the same unpriv user in the parent
namespace. This way I can create the needed
bind mounts in a mount namespace, and drop
the root privs. But only 1 extent is allowed, unless
you have CAP_SETUID in a parent namespace.

Does anyone (maybe @ebiederm or @amluto ?)
knows why it is so? I looked into `new_idmap_permitted()`
in a kernel, and I think it should be rather
straight-forward to change it to allow multiple
mappings to the same parent's uid. What was
the reason to only allow 1 mapping?

**Comment by ebiederm** [comment]
Stas Sergeev ***@***.***> writes:

> I started to experiment with user
> namespaces (which are needed together
> with mount namespaces here), and I
> don't understand why kernel allows to
> only write 1 extent (string) to `/proc/self/uid_map`.
> I would like to create 2 mappings: for root
> and unprivileged user within the namespace,
> mapping to the same unpriv user in the parent
> namespace. This way I can create the needed
> bind mounts in a mount namespace, and drop
> the root privs. But only 1 extent is allowed, unless
> you have CAP_SETUID in a parent namespace.
>
> Does anyone (maybe @ebiederm or @amluto ?)
> knows why it is so? I looked into `new_idmap_permitted()`
> in a kernel, and I think it should be rather
> straight-forward to change it to allow multiple
> mappings to the same parent's uid. What was
> the reason to only allow 1 mapping?

Because the mappings are 1-1.  Meaning you can't have the same uid
mapped twice in either direction.

It doesn't matter for setting up bind mounts.  The initial user in a
user namespace has all caps (within their user namespace).  Which means
in practice the mappings only matter for stat, and exec.

So what I would do is identity map the uid I start with, setup bind
mounts and then drop caps.

Eric

**Comment by stsp** [comment]
Ahha, thanks!
Very interesting.

Having the multiple mappings to the
same UID can help porting the existing
programs (which is the case here), because
then the caps are dropped on euid change
from 0 to non-0. dosemu had such a code
from the very beginning.
Of course if such possibility is not planned
for the kernel, then I'll have to add a new
code in parallel with old one, that would
use libcap.

**Comment by stsp** [comment]
Is there some library to automate the
chroot env creation? I need to bind-mount
too many things to create a full-blown
chroot, so maybe there is some automation
of that task? Some `libbubblewrap` could
help, but it seems there is none.

**Comment by stsp** [comment]
Ideally I don't want to do chroot.
All I want is a namespace with user's
files "hidden" from my process.
Is there some way to do that?
Or should the full chroot be created
just for that? :(
Can someone please guide me to
the right direction?

**Comment by stsp** [comment]
I think I just need to install the
seccomp filter to disallow the
open()s of a file not belonging
to dosemu2 group (aka group-based
solution on steroids). And make
that optional...

**Comment by stsp** [comment]
There appears to be a completely new
sandboxing framework called landlock:
https://docs.kernel.org/userspace-api/landlock.html
Seems to be much better for our task
than the namespace-based approach,
as it doesn't require creating chroot env.

**Comment by stsp** [comment]
> Regular users can’t give files away like that

Indeed, not like that.
Here's what seems to work:
`setfacl -m u:dosemu2:rw file`.
This is an unprivileged operation,
and I suppose dosemu can just
put the needed ACLs on its own,
and then setuid/setgid to dosemu2.
So I am back to the initial idea of
an UID-based separation, after all.
Hand-crafted sandbox is a dirty
business and people will find 1024+
ways of escaping it, because I've
never written a sandboxes in the
past, and getting it right from the
first time is simply not possible.
UID-based separation is OTOH
absolutely straight-forward.

In this scenario, the only priv setup
is done upon `sudo make install`.
You need to create dosemu2:dosemu2
user:group, chows the binary to that
and add suid/sgid bits.
On start, dosemu will add the needed
ACLs to the files and then switch UIDs.

Sounds ok?
I only wonder how difficult is that for
the packaging system to propagate.
Does debian packaging infrastructure
supports adding users and remapping
owner UIDs on a files, since the numeric
UIDs may not match?
If it is a problem for packaging, then
the sudo-based setup is possible, but
in this case someone (either user or
an installer) needs to modify /etc/sudoers.

**Comment by jschwartzenberg** [comment]
Is there a particular reason to have this for just dosemu? It would seem fitting to use Quebes OS or have another mechanism to separate _all_ programs from each other. I think Snap also has functionality for this to hide a user's regular files. Things that need to be available for a particular Snap need to be inside it's directory below `~/snap`.

**Comment by stsp** [comment]
dosemu runs untrusted binaries,
and therefore needs to establish
the sandboxing by itself too.
It doesn't care about separating
all programs from each other. It
cares about separating the untrusted
binaries from the host.

**Comment by jschwartzenberg** [comment]
Possibly the easiest way to achieve this would be to configure dosemu to use an image and not let it access any filesystem.

**Comment by stsp** [comment]
The configuration is irrelevant.
If dosemu2 is hacked, the attacker
gains all its abilities, not only those
configured for DOS progs.

**Comment by stsp** [comment]
Also note that configuration is already
secure: there is no way to mount arbitrary
dir any longer (was possible in dosemu1).
So restricting to hdimages doesn't help
a tiny bit.

**Comment by jschwartzenberg** [comment]
Can you give an example of an attack scenario that you're trying to address with this ticket? Malicious dos binaries that break out of the dosemu sandbox?

**Comment by stsp** [comment]
Exactly.

**Comment by jschwartzenberg** [comment]
How do you see this compared to the current implementations of for example QEMU and DOSBox? I would say it would be unfortunate if each application would reinvent the wheel on its own here. Maybe there are other similar tools that have something that would seem suitable to look at.

**Comment by stsp** [comment]
https://qemu-project.gitlab.io/qemu/system/security.html
Please read at least the
https://qemu-project.gitlab.io/qemu/system/security.html#principle-of-least-privilege
and
https://qemu-project.gitlab.io/qemu/system/security.html#isolation-mechanisms
sections.

dosbox seems to only have the
`-securemode` switch that restricts
mounts - not the security model I'd
like to have in dosemu2: we already
restrict mounts w/o any switches.
Security cannot be "optional".

**Comment by stsp** [comment]
Note that QEMU doesn't even list
the ACL+UID separation approach
that I am considering. It lists the
heavy-weight sandboxing approaches
(seccomp, namespaces) that I don't
want to deal with. If they have enough
of resources and skills to do it that
way - fine. I have neither the
resources nor skills for a full-blown
sandbox impl, and I've yet to understand
the disadvantages of my idea compared
to those.

**Comment by stsp** [comment]
Hi @ebiederm 

I am currently playing with the kernel
patch that adds a new `openat2()` flag
which I called `OA2_INHERIT_CRED`.
This flag allows to use fsuid/fsgid that
were in effect when dirfd was opened.
This helps me with the aforementioned
UID-based separation technique, because
when I setuid() to dosemu2, I normally
can't create the files owned by the original
user.
To overcome this, I can open dirfd
before setuid() and then use the
`OA2_INHERIT_CRED` flag (which should
only work with restricted path-resolution
modes) to create the files owned by the
original user under that particular dir.

Since it would be you anyway to review
that patch, could you please tell me
beforehand if it is an acceptable idea?
If not, I'll save some time by not posting
that patch to lkml.

**Comment by stsp** [comment]
So the UID separation sandbox doesn't
work because of this recent GTK restriction:
https://gitlab.gnome.org/GNOME/gtk/-/blob/main/gtk/gtkmain.c#L231
I filled the bug report:
https://gitlab.gnome.org/GNOME/gtk/-/issues/6629
but I don't think anything will come
out of that...

**Comment by stsp** [comment]
Filled 2 bugs for shadow-utils:
https://github.com/shadow-maint/shadow/issues/981
https://github.com/shadow-maint/shadow/issues/982

In a mean time I started implementing that
sanbox technique in a "suid" branch:
https://github.com/dosemu2/dosemu2/tree/suid
I still think this is simpler than namespace-based
solution, but already 3 external projects must be
patched (including linux kernel), so maybe this idea
is doomed.

**Comment by stsp** [comment]
Next problem is that `fsetxattr()`
does an (apparently undocumented)
fsuid check, so with an altered euid it
doesn't work. This however is documented
in `acl_set_fd()` which uses `fsetxattr()`.
So neither xattrs nor acls are going to work...
To get past this, I'd need to hack over
`setxattrat()`, which seems to be not
even upstreamed yet:
https://lore.kernel.org/lkml/20230511150802.737477-1-cgzones@googlemail.com/

**Comment by stsp** [comment]
Filled glibc rfe:
https://sourceware.org/bugzilla/show_bug.cgi?id=31664
In a mean time, in suid branch
`dosemu -X` mostly works (with
the still unpublished kernel patches).
Without `-X` doesn't even start. :(

**Comment by stsp** [comment]
So I published the first patch-set:
https://lore.kernel.org/lkml/20240422084505.3465238-2-stsp2@yandex.ru/
and a changes were requested.

While we are at it, @stefand could
you please tell me if I can advertise
this as being useful for wine as well?

In particular, after reading this:
https://wiki.archlinux.org/title/wine#Running_Wine_under_a_separate_user_account
I am under an impression that wine
seeks for such a solution too.

Advertising something as useful for
wine, is an exceptionally huge booster.
So if you tell me this is useful for wine,
I'll get that stuff in with nearly the 100%
certainty, whereas no one cares about
dosemu2-specific tricks.

**Comment by stefand** [comment]
Hi @stsp, long time no see :-)

I don't think Wine would use such functionality anytime soon. We are very explicit that Wine is not intended as a sandbox and that if a user wants to protect themselves against potentially malicious Windows software, they need to use other Linux tools to sandbox that software. I.e., you should treat wine+untrusted windows app as an untrusted Linux app.

Wine's API surface area is a lot larger than DOS, so adding sandbox functionality to Wine itself would be a lot more complicated. We don't have the developer resources to even attempt it.

**Comment by stsp** [comment]
Ok, thank you. :)
I hope eventually you'll re-evaluate
this vision, because what I propose
is very simple and in fact, is mostly
similar to what's there is already in
a howto linked above: you just run
wine as a different user! And this is
all of a sandboxing that I propose.
I only add the needed "tooling" to
make this transparently for the user,
rather than writing a howtos the way
you currently do.

But for now I won't advertise this as
being useful for wine.

**Comment by stefand** [comment]
As for features we made a very deliberate choice not to implement: Unlike DOS, we're running after a moving target. If Microsoft stops adding new things to Windows and we have another 30 years to catch up then we can reevaluate :-)

**Comment by stsp** [comment]
:)
Its not always happening that way.
Like many years ago you were telling
me the same about no-multilib, and
suddenly its there just because some
pressure was applied (when macos
dropped multilib).
So never say never. :)

**Comment by stefand** [comment]
Yeah, but Windows-Style multilib was a multi-year effort that involved full time work by a handful of devs. There's a plethora of other things that were not implemented because of it.

Plus the Mac side had a big carrot in addition to its stick: Codeweavers makes lots of money from Mac customers and Apple was providing a lot of help (Allowing 32 bit code segments, 32 bit support in Rosetta 2).

If someone needs Wine to become a user-friendly security sandbox, implementing this in Wine (as opposed to layering with external software) turns out to be a technically sound choice and that someone pays a million dollars to have this feature developed, then yes, we or someone else will probably do it.

**Comment by jschwartzenberg** [comment]
Maybe this thread is also interesting:
https://mastodon.social/@pid_eins/112353379708741872

**Comment by stsp** [comment]
> Maybe this thread is also interesting:

Yeah, many people want to remove suid
bits, its not something new. And only I
want to use them as a base for a sandboxing
technique. :)
When you are against the entire world,
the thing is probably not going to happen,
but at least with my last version of patches,
no one at lkml have yet pointed to a security
hole (first ones were buggy for sure).

**Comment by stsp** [comment]
I think if the patches are not accepted,
it would be still possible to implement
the same thing using a forked file server
process, as @amluto have suggested.
Much more work, but same result, same
suid-based model.

**Comment by stsp** [comment]
So on this branch, the ndmake error
persists, but I still can't reproduce it
locally.
Andrew, maybe you can on this branch?

**Comment by andrewbird** [comment]
I'll have a try.

**Comment by andrewbird** [comment]
PCMOS build ran fine here, no nmake error. I did see this later when just starting up dosemu
```backtrace
ajb@calypso:/clients/common/dosemu2.git$ ./runstandalonetest.sh 
fsrpc_svc1.c:210:20: runtime error: call to function _add_path_1 through pointer to incorrect function type 'int (*)(char *, void *, struct svc_req *)'
/clients/common/dosemu2.git/src/base/lib/fsrpc/fsrpc_svc1.c:23:25: note: _add_path_1 defined here
    #0 0x62d920e5ca11 in fsrpc_1 /clients/common/dosemu2.git/src/base/lib/fsrpc/fsrpc_svc1.c:210:20
    #1 0x77ee0df01170 in svc_getreq_common (/lib/x86_64-linux-gnu/libtirpc.so.3+0x1f170) (BuildId: e60d97c0a2c528644a7be9a76ffb6b1cf590e141)
    #2 0x77ee0df01345 in svc_getreq_poll (/lib/x86_64-linux-gnu/libtirpc.so.3+0x1f345) (BuildId: e60d97c0a2c528644a7be9a76ffb6b1cf590e141)
    #3 0x77ee0df013e1 in svc_run (/lib/x86_64-linux-gnu/libtirpc.so.3+0x1f3e1) (BuildId: e60d97c0a2c528644a7be9a76ffb6b1cf590e141)
    #4 0x62d920e5ca6d in fsrpc_svc_run /clients/common/dosemu2.git/src/base/lib/fsrpc/fsrpc_svc1.c:257:2
    #5 0x62d920e5ec5c in fssvc_init /clients/common/dosemu2.git/src/base/lib/fsrpc/fssvc.c:76:13
    #6 0x62d920ee489b in mfs_priv_init /clients/common/dosemu2.git/src/dosext/mfs/mfs.c:868:13
    #7 0x62d920df7ac5 in main /clients/common/dosemu2.git/src/base/core/emu.c:296:5
    #8 0x77ee0dc2a1c9 in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #9 0x77ee0dc2a28a in __libc_start_main csu/../csu/libc-start.c:360:3
    #10 0x62d920d667e4 in _start (/clients/common/dosemu2.git/2.0-pre9/bin/dosemu.bin+0x1cc7e4) (BuildId: b2917cfae7c7f99a32f96d7d3405fd2d5365f7d3)

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior fsrpc_svc1.c:210:20 
```

**Comment by stsp** [comment]
Is this a clang or gcc build?
Does it work correctly after an error?

**Comment by andrewbird** [comment]
clang build, is that correct? Seems to work okay, though I didn't try much more than dir. BTW my filesystem is NFS, not sure if that has an effect.

**Comment by stsp** [comment]
I don't know if this is correct or not.
But:
- I have no such error with gcc+ubsan
- this code is auto-generated, so I can't improve it

**Comment by andrewbird** [comment]
I did see some warnings during build, but it didn't look significant.
```make
fsrpc_srv.c:70:45: warning: omitting the parameter name in a function definition is a C23 extension [-Wc23-extensions]
   70 | bool_t seal_1_svc(int *ret, struct svc_req *)
      |                                             ^
fsrpc_srv.c:154:25: warning: omitting the parameter name in a function definition is a C23 extension [-Wc23-extensions]
  154 |         struct svc_req *)
      |                         ^
fsrpc_srv.c:167:25: warning: omitting the parameter name in a function definition is a C23 extension [-Wc23-extensions]
  167 |         struct svc_req *)
      |                         ^
fsrpc_srv.c:180:77: warning: omitting the parameter name in a function definition is a C23 extension [-Wc23-extensions]
  180 | bool_t unlink_1_svc(int idx, char *path, struct ret_s *ret, struct svc_req *)
      |                                                                             ^
fsrpc_srv.c:188:25: warning: omitting the parameter name in a function definition is a C23 extension [-Wc23-extensions]
  188 |         struct svc_req *)
      |                         ^
fsrpc_srv.c:195:79: warning: omitting the parameter name in a function definition is a C23 extension [-Wc23-extensions]
  195 | bool_t getxattr_1_svc(int idx, char *path, struct ret_s *ret, struct svc_req *)
      |                                                                               ^
fsrpc_srv.c:203:44: warning: omitting the parameter name in a function definition is a C23 extension [-Wc23-extensions]
  203 |         struct ret_s *ret, struct svc_req *)
      |                                            ^
fsrpc_srv.c:212:25: warning: omitting the parameter name in a function definition is a C23 extension [-Wc23-extensions]
  212 |         struct svc_req *)
      |                         ^
fsrpc_srv.c:219:76: warning: omitting the parameter name in a function definition is a C23 extension [-Wc23-extensions]
  219 | bool_t rmdir_1_svc(int idx, char *path, struct ret_s *ret, struct svc_req *)
      |                                                                            ^
fsrpc_srv.c:226:69: warning: omitting the parameter name in a function definition is a C23 extension [-Wc23-extensions]
  226 | bool_t path_ok_1_svc(int idx, char *path, int *ret, struct svc_req *)
      |                                                                     ^
fsrpc_srv.c:232:46: warning: omitting the parameter name in a function definition is a C23 extension [-Wc23-extensions]
  232 | bool_t exit_1_svc(void *ret, struct svc_req *)
      |                                              ^
```

**Comment by stsp** [comment]
These should now be fixed.

**Comment by stsp** [comment]
Filled this ticket:
https://github.com/thkukuk/rpcsvc-proto/issues/18
But I really don't know if it is an
UB or not, because gcc doesn't complain.

**Comment by stsp** [comment]
Filled this ticket:
https://savannah.gnu.org/bugs/index.php?65737
Now even GNU make resists to
this change-set...

**Comment by stsp** [comment]
Disabled clang's sanitizer to avoid
the problem. Not sure how good
such solution is, but for now this is
all I have.
This whole change-set feels a bit
unlucky, it seems.

**Comment by andrewbird** [comment]
So I retested with your latest suid branch there were no warnings during build or at startup, and it successfully built pcmos without error.

**Comment by stsp** [comment]
Ok, thanks, so still no way to
reproduce the instability locally. :(

**Comment by stsp** [comment]
They say rpcgen is deprecated.
So unless someone can point me
to something better, I am going
to look into searpc.

**Comment by stuaxo** [comment]
I used zeromq with msgpack to encode data for comms between different parts of a system before and it wasn't too bad - no sure if it's suitable for this.

zeromq handles making sure enough data gets where you want, then msgpack was relatively lightweight.

Having said that searpc looks pretty suitable.

**Comment by stsp** [comment]
I pushed the changes.
Lets see how it all goes...

**Comment by stsp** [comment]
So this security model is actually
flawed. Even though I can change
the user and the group, I can't quit
the supplementary groups. So I have
to use CAP_SETGID, which is entirely
crazy (I take more privs instead of taking
less).

I've seen a few attempts at LKML
in solving that, none seems succeeded.
I think the simplest solution can be
to add a new PFA_GRP_RESTRICT flag,
like PFA_NO_NEW_PRIVS (can be set,
never cleared). That flag, once set,
will change the group perm check this
way:
- if process is not in the desired group, proceed to Others (as before)
- if group access denied, then deny (as before)
- if group access allowed, then ignore and proceed to Others (this is a new rule!)

In other words, both Group and Others
bits must be set in order for the access to
be granted.
This differs from other LKML solutions by
the restrictive manner. While dropping the
group list may be both restrictive and
permissive, the aforementioned trick
can't be permissive, it only restricts. And it
looks very simple to implement, too.

@amluto @ebiederm what do you think?
Can something like this work?

**Comment by stuaxo** [comment]
Maybe not useful here - but I'm reminded of Justine Tunnys go at porting BSD Pledge to Linux

https://justine.lol/pledge/

**Comment by ebiederm** [comment]
Stas Sergeev ***@***.***> writes:

> So this security model is actually
> flawed. Even though I can change
> the user and the group, I can't quit
> the supplementary groups. So I have
> to use CAP_SETGID, which is entirely
> crazy (I take more privs instead of taking
> less).
>
> I've seen a few attempts at LKML
> in solving that, none seems succeeded.
> I think the simplest solution can be
> to add a new PFA_GRP_RESTRICT flag,
> like PFA_NO_NEW_PRIVS (can be set,
> never cleared). That flag, once set,
> will change the group perm check this
> way:
> - if process is not in the desired group, proceed to Others (as before)
> - if group access denied, then deny (as before)
> - if group access allowed, then ignore and proceed to Others (this is a new rule!)
>
> In other words, both Group and Others
> bits must be set in order for the access to
> be granted.
> This differs from other LKML solutions by
> the restrictive manner. While dropping the
> group list may be both restrictive and
> permissive, the aforementioned trick
> can't be permissive, it only restricts. And it
> looks very simple to implement, too.
>
> @amluto @ebiederm what do you think?
> Can something like this work?

There exists uidmap and gidmap that are very small binaries that can be
audited that can change things, according to what is in /etc/subuid and
/etc/subgid.

That could allow you to run dosemu as a uid and gid that does not
overlap with any other user on the system, while at the same time
gaining enough privileges (inside a user namespace) to change how things
are mounted to reduce your visibility of the system even more.

That is already widely depolyed so you can take advantage of it now.


Of course any of that needs to address which libraries are linked in,
and whose constructors run before main is reached.  To avoid that attack
it probably makes most sense use an extra executable like bubblewrap
(that only needs to be linked to libc) to setup the environment for you.
Not that you have to use bubblewrap you can roll your own binary easily
enough.

Eric

**Comment by stsp** [comment]
Hi Eric, the lib constructors are not a
problem because dosemu2 (and in
fact, the latest versions of dosemu1,
too) uses dlopen() for all runtime deps.

The problem here is that I implemented
things not via the use of the namespaces,
for various reasons (eg I am not sure
how portable the namespaces are, etc).
What I do, is a pure classic: RPC server
that works as an original user, and the
main dosemu2 process switched to
dosemu2:dosemu2 via suid/sgid bits.
I am completely shocked that the techniques
that come from 80s, are not working
these days, as I can't quit the supplementary
groups.
This is why I propose this very simple
trick to restrict group perms with an
PFA_GRP_RESTRICT flag.

That should replace this proposal:
https://lore.kernel.org/all/0895c1f268bc0b01cc6c8ed4607d7c3953f49728.1416041823.git.josh@joshtriplett.org/
Yes, I realize that you provided the
replacement with namespaces already,
you said so in the aforementioned thread.
But this doesn't mean the original problem
have vanished.

**Comment by stsp** [comment]
Also I do really want to get rid of libcap,
as its entirely broken:
https://bugzilla.kernel.org/show_bug.cgi?id=219168
https://bugzilla.kernel.org/show_bug.cgi?id=219174

And, unless I am mistaken, your
namespace solution requires the
use of libcap/libpsx.
Why such a security-sensitive libs
like libcap, are that broken btw?

**Comment by ebiederm** [comment]
Stas Sergeev ***@***.***> writes:

> Hi Eric, the lib constructors are not a
> problem because dosemu2 (and in
> fact, the latest versions of dosemu1,
> too) uses dlopen() for all runtime deps.
>
> The problem here is that I implemented
> things not via the use of the namespaces,
> for various reasons (eg I am not sure
> how portable the namespaces are, etc).
> What I do, is a pure classic: RPC server
> that works as an original user, and the
> main dosemu2 process switched to
> dosemu2:dosemu2 via suid/sgid bits.
> I am completely shocked that the techniques
> that come from 80s, are not working
> these days, as I can't quit the supplementary
> groups.

My apologies then.  I had not realized the complaint was in the context
of just using suid and sgid bits.

Long story short it has never been possible to quit supplementary
groups.  Supplementary groups were designed so that they could include
negative groups that subtracted permissions from users.

AKA it has historically been possible to be added to the
dont-trust-these-processes group.

> This is why I propose this very simple
> trick to restrict group perms with an
> PFA_GRP_RESTRICT flag.

I see what you are proposing now and why.
I will send a more targeted reply to your original comment.

Eric

**Comment by stsp** [comment]
> I had not realized the complaint was in the context
> of just using suid and sgid bits.

Yeah, I thought if I take the simple classic
approach, then what could go wrong?
Except that everything went wrong.
Starting from broken gtk:
https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/7246
aforementioned libcap/libpsx breakage,
inability to integrate that scheme into
flatpak security model, and so on.
And finally it doesn't work at all (w/o
CAP_SETGID, which adds more problems
than solves) because of that supplementary
group problem.

> I see what you are proposing now and why.

This was a minimal proposal.
Maybe some day it would be good to consider
the mechanism of sending the supplementary
group list via the unix socket. In this case the
privileged helper (aka systemd) will do getgroups()
for the needed UID and send the result via the
unix socket to the recipient. This will allow to
accompany the uid/gid switch with the proper
supplementary list switch.
What do you think?
But for now the minimal proposal is certainly
sufficient.

**Comment by ebiederm** [comment]
Stas Sergeev ***@***.***> writes:

> So this security model is actually
> flawed. Even though I can change
> the user and the group, I can't quit
> the supplementary groups. So I have
> to use CAP_SETGID, which is entirely
> crazy (I take more privs instead of taking
> less).
>
> I've seen a few attempts at LKML
> in solving that, none seems succeeded.
> I think the simplest solution can be
> to add a new PFA_GRP_RESTRICT flag,
> like PFA_NO_NEW_PRIVS (can be set,
> never cleared). That flag, once set,
> will change the group perm check this
> way:
> - if process is not in the desired group, proceed to Others (as before)
> - if group access denied, then deny (as before)
> - if group access allowed, then ignore and proceed to Others (this is
> a new rule!)
>
> In other words, both Group and Others
> bits must be set in order for the access to
> be granted.
> This differs from other LKML solutions by
> the restrictive manner. While dropping the
> group list may be both restrictive and
> permissive, the aforementioned trick
> can't be permissive, it only restricts. And it
> looks very simple to implement, too.
>
> @amluto @ebiederm what do you think?
> Can something like this work?

I am not at all certain I understand the proposal.  Fundamentally as
long as you are a member of a group and that membership denies you
access setgroups can not allow you to drop groups.  You aren't making a
change that would allow setgroups to drop groups.

Perhaps you were thinking of a change that when the group matches and
permission are denied, treat it as if the group did not match.  And a
similar change for posix and nfs acls.

Let me call the case where a group match results in fewer permissions
than other a negative group.  Likewise a user match that results
in fewer permissions than other a negative user.

I think anything that sets permissions would need to ensure that
user and group permissions are not more restrictive than other
permissions, so that anyone who attempts to create a negative group
would quickly discover they can't be created.

Removing kernel support of negative groups, would allow setgroups to
safely drop groups.  Assuming that userspace isn't being more
restrictive if a process is in a specific group.

I think it can be done.  It would take some conversations and finding
and updating all of the relevant userspace tools.  But since negative
groups are quite rare I think it is doable.

The big question is what are the advantages to not having negative
groups.  If there is a significant advantage I can see people being
persuaded to break with history and remove support for negative groups.

I expect at the same time as removing support for negative groups we
should remove support for negative users as well.

Eric

**Comment by stsp** [comment]
> You aren't making a
> change that would allow setgroups to drop groups.

Yes, as that was already proposed by
others. My proposal is new and is different.

> Perhaps you were thinking of a change that when the group matches and
> permission are denied, treat it as if the group did not match.

The opposite: if the group matches and
permissions are ALLOWED (by the Group
bits), then treat as the group did not match.
I.e. require also Others allowed, together
with Group allowed. That's a pure restriction.

Removing kernel support for "negative"
groups may be doable, but definitely not
for dosemu2. Its a big decision, not me
being the right person to propose such.
Maybe a new personality flag can be used
for that, I've no idea.

What I propose is however very simple:
just a new PFA_xx flag to make the aforementioned
restriction. When the group matches and
permissions ALLOWED, proceed to Others.

**Comment by stsp** [comment]
> Maybe not useful here - but I'm reminded of Justine Tunnys go at porting BSD Pledge to Linux

At least pledge() is possible on top
of bpf, so this is not the real problem.
But when you ask yourself why you
still can't change uid/gid effectively
(no handling for supplementary groups),
or why noexec mount still breaks
shm_open() (shm_open() could be
a separate syscall rather than a wrapper
on top of open() for a fine-grained control),
or why libcap seeks the ways of performing
the cap operations in a process scope
rather than the thread scope and can't
find any, or why failed mmap()/mremap()
keeps the previously mapped regions
unmapped (i.e. failed syscall doesn't
restore the old state), why gtk doesn't
permit suid/sgid bits even if you want
to switch to non-root user, why fuse
doesn't allow fds passed to another
user's process to work w/o allow_other
mount option which itself requires root
to enable, and so on...
then you suddenly think that linux did
a great job in the namespaces area,
but the classic security approaches are
all in the dust.

**Comment by ebiederm** [comment]
Stas Sergeev ***@***.***> writes:

>> You aren't making a
>> change that would allow setgroups to drop groups.
>
> Yes, as that was already proposed by
> others. My proposal is new and is different.
>
>> Perhaps you were thinking of a change that when the group matches and
>> permission are denied, treat it as if the group did not match.
>
> The opposite: if the group matches and
> permissions are ALLOWED (by the Group
> bits), then treat as the group did not match.
> I.e. require also Others allowed, together
> with Group allowed. That's a pure restriction.
>
> Removing kernel support for "negative"
> groups may be doable, but definitely not
> for dosemu2. Its a big decision, not me
> being the right person to propose such.
> Maybe a new personality flag can be used
> for that, I've no idea.
>
> What I propose is however very simple:
> just a new PFA_xx flag to make the aforementioned
> restriction. When the group matches and
> permissions ALLOWED, proceed to Others.

My thoughts:

- What does restricting the group match to the permissions of the
  other match get you?

- In general it is desirable to make a user or a group match allow more
  than an other match, so I don't see how it would ever make sense to
  limit a group match to the permission of the other match.

  Negative groups we can probably get away with as I only know
  of one instance (over 30 years) where someone has used them.
  Positive groups are used all of the time for all kinds of things
  so I can't imagine crippling them not breaking lots of things.

- I think I saw a desire to start a worker process with absolutely
  minimal privileges (using suid and sgid).  In my experience changing
  permissions upon exec has too many knobs someone calling exec can
  fiddle with to make it a reliable way to start an executable.  The
  better way is to use a daemon (inetd and systemd are prebuilt daemons)
  to start a worker process.  At that point the work process starts up
  the same way every time and so you don't have to worry about what
  the clever thing the program calling exec did.

Eric

**Comment by ebiederm** [comment]
Stas Sergeev ***@***.***> writes:

> why noexec mount still breaks
> shm_open() (shm_open() could be
> a separate syscall rather than a wrapper
> on top of open() for a fine-grained control),

What is the problem with shm_open and noexec?
I haven't heard of that one.

Are you trying to do something such as create an executable file with
shm_open on a noexec mount?

Eric

**Comment by stsp** [comment]
> - What does restricting the group match to the permissions of the
>  other match get you?

After I switch uid/gid to dosemu2:dosemu2,
I expect that user to have no permissions at
all among the files of the original user.
However the supplementary group list, which
always includes the original GID too, removes
all the restrictions! dosemu2:dosemu2 process
can still access anything group-accessible for
the original user!
So I just need that new flag to restrict also that.

> - In general it is desirable to make a user or a group match allow more
>   than an other match, so I don't see how it would ever make sense to
>   limit a group match to the permission of the other match.

It only makes sense after the suid/sgid-assisted
switch. It is not needed for anything else.

> - I think I saw a desire to start a worker process with absolutely
>   minimal privileges (using suid and sgid).

Its somewhat the other way around
here. I spawn the worker process with
normal user privs and it servs as an
RPC server for passing in fds to me.
And then I lower the priv of the main.

> What is the problem with shm_open and noexec?
> I haven't heard of that one.

That you can't do mmap(PROT_EXEC) on any
files from the noexec-mounted fs. For things
like dosemu2 and many others, they initially
left a back-door: you could mmap() w/o
PROT_EXEC and then mremap(). But recently
that was treated as a security hole and "fixed",
leaving no work-around. Intentional compatibility
break! This besides the fact that the malicious
software can still use read() to the anonymously
mapped PROT_EXEC mapping, so this technique
only breaks the ligitimate progs, but doesn't fix
anything at all. I tried to fight that in LKML
something like 20 years ago and others tried as
well, but it was pushed by the famous Ulrich
Drepper, so no fighting was possible.
Note that making shm_open() a syscall can
just avoid the problem all together.
Is this a good security, really?

**Comment by stsp** [comment]
Btw, I appreciate that now, 20 years
late, the complains are at least listened. :)
I've got used to the fact that no one just
cares, and once you send the RFC patch
yourself, it is criticised for bugs and rejected.
In the very rare cases when @amluto stepped
in, I was at least getting some attention
and the useful suggestions. But otherwise
nothing.

**Comment by stsp** [comment]
> you could mmap() w/o
> PROT_EXEC and then mremap().

This reads as: "and then mprotect()".

**Comment by stsp** [comment]
> Note that making shm_open() a syscall can

To elaborate:
Such syscall can return an fd that bypasses
the PROT_EXEC restriction, but does not
permit read()/write() calls - only mmap().
Then add a new mount flag "noopen" to
completely suppress the open() calls.
Then the only way to work with such mounts
would be via this restricted shm_open() syscall.
Ideally I think shms should use per-user
dirs, not one global dir, for even better
resource separation. Which is what I do
now by hands, in /tmp. As /dev/shm is
broken, I had to hack up my own shm APIs. :(

**Comment by ebiederm** [comment]
Stas Sergeev ***@***.***> writes:

>> - What does restricting the group match to the permissions of the
>>  other match get you?
>
> After I switch uid/gid to dosemu2:dosemu2,
> I expect that user to have no permissions at
> all among the files of the original user.
> However the supplementary group list, which
> always includes the original GID too, removes
> all the restrictions! dosemu2:dosemu2 process
> can still access anything group-accessible for
> the original user!
> So I just need that new flag to restrict also that.

On the face of it that seems reasonable.

It would take a combination of bits that are used for nothing else,
or some security xattr.


OTOH I think it would need to be something like a security xattr
that only root cat set to make it workable.  Otherwise it would
provide a mechanism that ordinary users could use to drop negative
groups, and gain permissions.

I still think your best bet for security is a daemon that spawns dosemu
instances, but your proposal could work.

It is my firm belief that the suid bit (and other versions of changing
permissions on exec is an experiment that has failed).  There are so
many different kinds of state that can be passed through exec that it
seems to be a continual game of whack-a-mole to close prevent such
executables from being exploited.

>> What is the problem with shm_open and noexec?
>> I haven't heard of that one.
>
> That you can't do mmap(PROT_EXEC) on any
> files from the noexec-mounted fs. For things
> like dosemu2 and many others, they initially
> left a back-door: you could mmap() w/o
> PROT_EXEC and then mremap(). But recently
> that was treated as a security hole and "fixed",
> leaving no work-around. Intentional compatibility
> break! This besides the fact that the malicious
> software can still use read() to the anonymously
> mapped PROT_EXEC mapping, so this technique
> only breaks the ligitimate progs, but doesn't fix
> anything at all. I tried to fight that in LKML
> something like 20 years ago and others tried as
> well, but it was pushed by the famous Ulrich
> Drepper, so no fighting was possible.
> Note that making shm_open() a syscall can
> just avoid the problem all together.
> Is this a good security, really?

The restrictions do seem a bit silly.

One of the problems when dealing with security in the unix model is
often the people fixing the problems have information of actual problems
they choose not to share when fixing problems because it would make
those problems too easy to exploit.  That in turn messes up code review.
So it is possible there is a good reason for noexec being enforced.

I don't think making shm_open a proper system call would make the
situation fundamentally better than having a filesystem.  There is at
least the execat system call that can be used to execute the contents
of the file descriptor.

That said I don't currently see any situation where mremap/mprotect to
add PROT_EXEC is problematic.

Hmm....  I can imagine that the shared library loaders might mmap parts
of the file and the mprotect it to being executable if need be.  In
which case mprotect not honoring the filesystem permissions could be a
foot gun.  It is easy enough to prove of disprove by looking at the
dynamic loader.

All security fixes changing the ABI run the risk of breaking things.
I would have thought an actual regression report would have been enough
to keep that from happening.  It is a shame that didn't work at the time.

Eric

**Comment by stsp** [comment]
> OTOH I think it would need to be something like a security xattr
> that only root cat set to make it workable.

I am not quite sure what do you
mean here. You seem to imply enabling
that feature per-file, in xattrs?
My proposal has nothing to do with
per-file enablement. I want a per-process
flag, similar to PFA_NO_NEW_PRIVS, which
I can set with prctl() but never clear (even
on exec).
And since linux doesn't seem to have
the "per-process" concept but only
"per-thread", I guess this flag would
be per-thread, and I'll have troubles
setting it for every thread... like libcap
has problems setting caps for all threads.

> I don't think making shm_open a proper system call would make the
> situation fundamentally better than having a filesystem.

I am not proposing to get rid of tmpfs
filesystem. But if you have shm_open()
as a separate syscall, you can:
1. bypass "noexec" restriction with that syscall
2. add some "noopen" mount flag to completely forbid open() call (that, in turn, forbids exec-via-fd)
3. make that fd returned by shm_open() to disallow read()/write(), leaving only mmap()
That said, shm_open() syscall should still use
/dev/shm as before, but the advantage is the
arbitrary restrictions you can have on that fd.

> Hmm....  I can imagine that the shared library loaders might mmap parts
> of the file and the mprotect it to being executable if need be.

I think shared library loaders can change,
so looking at their code may not be exactly
useful. If there is a valid attack scenario, then
the fd returned by open() should not allow
mmap(PROT_EXEC), but fd returned by shm_open()
can still allow that. Of course since I don't know
that attack scenario, I am just speculating.
But I don't think they covered any, because,
fundamentally, you can still read() to executable
area. If you have a separate shm_open() syscall,
then you can completely forbid read()/write(),
so this "hole" is also closed.

**Comment by stsp** [comment]
> I still think your best bet for security is a daemon that spawns dosemu
> instances

Do you mean a suid-root daemon?
As I don't think you can run dosemu
as a different user w/o suid-root.
And unfortunately I think suid-root
helper is not an option, as people will
never buy yet another suid-root binary
added to their system.
I am currently using cap_setgid, which
is similarly as bad. I am sure if I won't
find the cap-less method quickly enough,
people will beat me for cap_setgid and
will start advocating to dump dosemu2
forever.

**Comment by stsp** [comment]
> It is my firm belief that the suid bit (and other versions of changing
> permissions on exec is an experiment that has failed).

You mean, even setcap?

>  There are so
> many different kinds of state that can be passed through exec that it
> seems to be a continual game of whack-a-mole to close prevent such
> executables from being exploited.

Indeed, but, unlike eg setcap, suid/sgid
bits may be relatively safe in case the
owner is non-root but rather some unpriv'd
user. So it really seems to be the only
sane way forward. And it looks like there
is just one bit (to restrict groups) needed
to turn suid/sgid bits from "useless crap"
into something really powerful. I suppose
this was also the Josh Tripplet's point when
he proposed to drop the group list. But my
proposal is safer as it doesn't remove the
negative matches, it only restricts positive
matches.

**Comment by stsp** [comment]
@ebiederm 
Actually your xattr mentioning makes me think
of a much simpler solution to the noexec problem:
`O_CREAT | O_EXCL` combo gives you an fd that
can be used to set the special security xattr that
allows to bypass the `noexec` restriction for the
file owners (by UID).
I.e. only if you have created this file, or you
have CAP_SYS_ADMIN, then you can allow
the file owners to bypass the `noexec` restriction.

Can something like this work?
Hmm, this makes me think of the need to have
a way to provide xattrs directly to the open(O_CREAT)
call, together with mode bits... Or is there such
technique already?

**Comment by stsp** [comment]
Posted the LKML proposal:
https://lore.kernel.org/lkml/20240930063405.113227-1-stsp2@yandex.ru/T/#u

**Comment by stsp** [comment]
There weren't too many comments,
just one change request. The final version
of patch is here:
https://www.spinics.net/lists/kernel/msg5383847.html
@amluto @ebiederm seems like the
things won't go without your ACK or
comments.

**Comment by stsp** [comment]
@amluto @ebiederm 
Is there any way to get the same
effect of switching the user, but
w/o having to add the user? Container
people complain to the need of
adding a user. So how can I disable
the User and Group access? Should
I look into BPF or Landlock or some
other newfangled technique?

**Comment by stsp** [comment]
https://lore.kernel.org/lkml/20241107215821.1514623-2-stsp2@yandex.ru/T/
New proposal posted.
This one is supposed to actually
solve the problem by having the
correct group list at the end.

**Comment by jschwartzenberg** [comment]
Maybe this project is also interesting: https://github.com/Zouuup/landrun

**Comment by stsp** [comment]
No because:
```
Linux Landlock is a kernel-native security module that lets unprivileged processes sandbox themselves - but
nobody uses it because the API is ... hard!
```
Nobody, really?
Quite a bold statement, I use
Landlock directly in dosemu2,
and its a fantastic framework
indeed. Unfortunately it doesn't
cover all of the dosemu2 needs.
They currently work on a
per-process landlocking (rather
than per-thread) and a few
other things that I'll use, but still
the landlock model is insufficient
here. Plus the development
seems to have mostly stalled,
so I am not sure if even the
already planned extensions
are going to be implemented. :(

Currently dosemu2 is trying to
glue 1024+ different security
techniques together (like landlock,
libcap, libacl and so on), which
is not a way to go. But on a positive
side, dosemu2 is now split into
multiple security domains with
libSeaRPC.

**Comment by stuaxo** [comment]
Partially relevant: but I wish Linux would pick up the BSD Pledge API, it seems a lot easier to understand than some of APIs in Linux for security.

**Comment by stsp** [comment]
Here you go:
https://github.com/jart/pledge
But I find this completely useless
for dosemu2. And I'd really like
to see who uses that effectively,
how and why.
